### PR TITLE
Public STR and TB fields for verification

### PR DESCRIPTION
**Pull request by arlolra**
Reopening for discussion here since I messed up #34

### CODE REVIEW

File: merkletree/str.go [hunk 1]
```diff
@@ -79,3 +79,15 @@ func (str *SignedTreeRoot) PreviousSTRHash() []byte {
 func (str *SignedTreeRoot) Signature() []byte {
 	return str.sig
 }
+
+func (str *SignedTreeRoot) LibVersion() []byte {
```
Comment (masomel) [inline merkletree/str.go]: Shouldn't ``LibVersion()``, ``HashID()`` and ``EpochDeadlinePolicy()`` be in ``policy.go``?

File: merkletree/pad.go [hunk 2]
```diff
@@ -122,6 +122,7 @@ func (pad *PAD) GetSTR(epoch uint64) *SignedTreeRoot {
 func (pad *PAD) TB(key string, value []byte) (*TemporaryBinding, error) {
 	str := pad.latestSTR
 	index, _ := pad.computePrivateIndex(key, pad.policies.vrfPrivate())
+
```
Comment (masomel) [inline merkletree/pad.go]: I think we either want to keep the VRF proof here, or return the TB with a proof of absence as I mentioned in https://github.com/coniks-sys/coniks-go/pull/29#discussion_r72140992, no?

File: merkletree/tb.go [hunk 3]
```diff
@@ -1,6 +1,7 @@
 package merkletree
 
 type TemporaryBinding struct {
+	str   []byte
```
Comment (masomel) [inline merkletree/tb.go]: I'm wondering if we need to include a field for the STR signature in the TB if the STR struct also exposes the signature, and the server needs to return the STR to the client for verification anyway?

**Comment by masomel** [comment]
Alternatively to having public fields in the TB and STR, what I had in mind when I opened #24 was to expose a ``Verify()`` function. What do you think of 9480b16?

**Comment by masomel** [comment]
In my branch, I decided to leave the public fields that were added in 812caad since I'm not sure if the client will still need to access those fields using the getters.

File: merkletree/pad.go (hunk 2 above)
Comment (c633) [inline merkletree/pad.go]: Yeah. That's why I said that this branch should be merged after `vrf_integration` branch. Sorry for the noise :(

File: merkletree/str.go (hunk 1 above)
Comment (c633) [inline merkletree/str.go]: These can change in epochs, thus I think it should be here, corresponding with policies of a specific STR. How does that sound?

File: merkletree/tb.go (hunk 3 above)
Comment (c633) [inline merkletree/tb.go]: Yes. I just realize that the server will return STR proof, VRF proof, proof of absence and TB to the client. So this field should be removed.

**Comment by c633** [comment]
> Alternatively to having public fields in the TB and STR, what I had in mind when I opened #24 was to expose a Verify() function. What do you think of db24bfa?

I got your idea. But the signing key cannot be used here, since I presume that the `Verify()` function should be called in the client-side only. 

> In my branch, I decided to leave the public fields that were added in 812caad since I'm not sure if the client will still need to access those fields using the getters.

Actually, I think the key server would use those getters to return the STR/TB to the client. The client doesn't use those getters directly.

**Comment by masomel** [comment]
> the signing key cannot be used here, since I presume that the Verify() function should be called in the client-side only. 

Yes, the idea would be for the client to call ``Verify()``. I chose to use the ``SigningKey`` since that's what ``crypto.Verify()`` uses as well. If ``SigningKey`` is the server's key pair, I agree that ``Verify()`` shouldn't take it as a param, but then neither should the verification function in ``crypto``.

> Actually, I think the key server would use those getters to return the STR/TB to the client. The client doesn't use those getters directly. 

Ok, I see.

**Comment by c633** [comment]
> Yes, the idea would be for the client to call Verify(). I chose to use the SigningKey since that's what crypto.Verify() uses as well. If SigningKey is the server's key pair, I agree that Verify() shouldn't take it as a param, but then neither should the verification function in crypto.

Sorry for that. I removed it in 3da95ef40a19094ad9f95f556676649c1deb2daf

**Comment by masomel** [comment]
> Sorry for that. I removed it in 3da95ef

Yeah, with all the signature verification wrappers in place, I suppose we no longer need to export ``crypto``'s ``verify()``. I still think we want this function, to allow for the wrappers I created to work. I gave this a shot in d46c806. Please let me know what you think!

**Comment by masomel** [comment]
> I suppose we no longer need to export ``crypto``'s ``verify()``. I don't think we need to remove it entirely if we want something like ``VerifySTR()`` to work.

Argh, this was silly of me. I forgot ``crypto`` is no longer a subpackage of ``merkletree``. Sorry for the noise.

**Comment by masomel** [comment]
Like we've decided to keep the ``VerifyX()`` functions in the lib as we discussed in https://github.com/coniks-sys/coniks-go/commit/d46c806796a4f58b038fc1ee8794b2bccfd8e8aa#commitcomment-18452858, so I'll go ahead and merge in the ``str-tb-verify-idea`` branch and close this PR.

**Pull request closed without merging**